### DISCUSSION TITLE
How to automatically download trusted root certificate authority on Windows

### DISCUSSION THREAD

**Discussion by eric-324723** [curl/curl#14771] [Q&A] [answered]
There is a difference in behavior with this curl, and the curl natively included in Windows.

When making a request with Windows native curl, if a trusted root certificate authority such as one listed [here](https://learn.microsoft.com/en-us/security/trusted-root/participants-list) is not found in the local store, the request will still succeed and the certificate will be downloaded and added to the local store (accessed with certlm.msc)

When making the same request with the open-source curl.exe, and it is using the native store (--ca-native), the request will FAIL if the root certificate authority is not already in there.

How do I replicate this Windows native curl.exe behavior?

**Comment by dfandrich** [comment]
What TLS back-end is your curl.exe using? That behaviour might be an Schannel-specific thing.

**Reply by eric-324723** [reply]
`curl.exe --version`

```
curl 7.84.0 (x86_64-pc-win32) libcurl/7.84.0 OpenSSL/1.1.1d (Schannel) zlib/1.2.11 WinIDN nghttp2/1.44.0
Release-Date: 2022-06-27
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtsp smb smbs smtp smtps telnet tftp
Features: alt-svc AsynchDNS HSTS HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz MultiSSL NTLM SPNEGO SSL SSPI threadsafe UnixSockets
```

**Comment by jay** [comment, marked as answer]
> When making a request with Windows native curl, if a trusted root certificate authority such as one listed [here](https://learn.microsoft.com/en-us/security/trusted-root/participants-list) is not found in the local store, the request will still succeed and the certificate will be downloaded and added to the local store (accessed with certlm.msc)
> 
> When making the same request with the open-source curl.exe, and it is using the native store (--ca-native), the request will FAIL if the root certificate authority is not already in there.

See #12303. curl imports the CA certificates in the Windows root store but can't import what's not there. Windows does seem to retrieve missing root certificates on demand if Schannel is used for verification. I don't know what we can do about this because there is no documented way to do it Windows without using Schannel for verification.
